keras_nlp/layers/random_swap.py

```diff
@@ -0,0 +1,265 @@
+                    tf.convert_to_tensor(self.skip_list),
+                    tf.convert_to_tensor([True] * len(self.skip_list)),
+                ),
+                default_value=False,
+            )
+
+    @tf.function
+    def call(self, inputs):
+
+        if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+            inputs = tf.convert_to_tensor(inputs)
+
+        input_is_1d = False
+        if inputs.shape.rank < 1 or inputs.shape.rank > 2:
+            raise ValueError(
+                "Input must either be rank 1 or rank 2. Received input with "
+                f"rank={inputs.shape.rank}"
+            )
+        elif inputs.shape.rank == 1:
+            input_is_1d = True
+            # Add a new axis at the beginning.
+            inputs = tf.expand_dims(inputs, axis=0)
+        if isinstance(inputs, tf.Tensor):
+            # Convert to ragged tensor.
+            inputs = tf.RaggedTensor.from_tensor(inputs)
+
+        row_splits = inputs.row_splits
+
+        def _check_skip(token):
+            if self.skip_list:
+                return self.StaticHashTable.lookup(token)
+            elif self.skip_fn:
+                return self.skip_fn(token)
+            elif self.skip_py_fn:
+
+                def string_fn(token):
+                    return self.skip_py_fn(token.numpy().decode("utf-8"))
+
+                def int_fn(token):
+                    return self.skip_py_fn(token.numpy())
+
+                py_fn = string_fn if inputs.dtype == tf.string else int_fn
+
+                return tf.py_function(py_fn, [token], tf.bool)
+            else:
+                return False
+
+        positions_flat = tf.range(tf.size(inputs.flat_values))
+        positions = inputs.with_flat_values(positions_flat)
+        # Figure out how many we are going to select.
+        token_counts = tf.cast(inputs.row_lengths(), "float32")
+        num_to_select = tf.random.stateless_binomial(
+            shape=tf.shape(token_counts),
+            seed=self._generator.make_seeds()[:, 0],
+            counts=token_counts,
+            probs=self.rate,
+        )
+        if self.max_swaps is not None:
+            num_to_select = tf.math.minimum(num_to_select, self.max_swaps)
+        num_to_select = tf.cast(num_to_select, "int64")
+
+        def _swap(x):
+            positions, inputs, num_to_select = x
+            if tf.size(positions) == 1:
+                return positions
+            for _ in range(num_to_select):
+                index = tf.random.stateless_uniform(
+                    shape=tf.shape(positions),
+                    minval=0,
+                    maxval=tf.size(positions),
+                    dtype=tf.int32,
+                    seed=self._generator.make_seeds()[:, 0],
+                )
+                index1, index2 = index[0], index[1]
+                word1 = inputs[index1]
+                word2 = inputs[index2]
+                if _check_skip(word1) or _check_skip(word2):
```

Comment:
I'm unsure we should do the skipping this way. Probably better to first filter down to a list of candidate positions (using skip_fn), then use rate/max to determine the number of swaps, then apply the swaps to the candidate positions.

Similar to how we do this for deletions.